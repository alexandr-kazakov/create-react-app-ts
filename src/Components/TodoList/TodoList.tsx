import { useEffect, useState } from "react";
import Button from "../Button/Button";


type Todo = {
    id: number;
    title: string;
}

const apiUrl: string = 'https://jsonplaceholder.typicode.com/todos';

const TodoList: React.FC = () => {
    const [todos, setTodos] = useState<Todo[]>([]);
    const [inputValue, setInputValue] = useState<string>('');


    useEffect(() => {
        const storedTodosString = localStorage.getItem('todos');
        if (storedTodosString !== null) {
            const storedTodos = JSON.parse(storedTodosString);

            if (storedTodos && storedTodos.length > 0) {
                setTodos(storedTodos);
            } else {
                fetch(apiUrl)
                .then(response => response.json())
                .then(data => setTodos(data.slice(0, 10)))
                .catch(error => console.error('Error fetching data:', error));
            }
        }
    }, []);

    useEffect(() => {
        if (todos.length > 0) {
            localStorage.setItem('todos', JSON.stringify(todos));
        }
    }, [todos]);
      

    const handleAddTodo = () => {
        if(inputValue.trim() !== '') {
            setTodos([...todos, {id: todos.length + 1, title: inputValue}]);
            setInputValue('');
        }
    }

    const handleDeleteTodo = (id: number) => {
   
        const updatedTodos = todos.filter(todo => todo.id !== id);

        if(!updatedTodos.length) {
            console.log('todos', todos)
            localStorage.removeItem('todos');
        }
        setTodos(updatedTodos);
    }

    const handleClearTodos = () => {
        localStorage.removeItem('todos');
        setTodos([]);
    }

    return (
        <div className="todo-list">
            <div className="mb-10">
                <input className="dark:bg-black dark:text-white border border-slate-400 h-12 pl-2 pr-2" type="text" value={inputValue} onChange={event => setInputValue(event.target.value)} placeholder="Enter your task"></input>
                <Button onClick={handleAddTodo} className="bg-primary text-white pl-8 pr-8 pt-3 pb-3" text="Add"></Button>
            </div>

            <table className="table-auto border-collapse border border-slate-400 max-w-5xl w-full mb-10">
                <thead>
                    <tr>
                    <th className="border border-slate-300">Info</th>
                    <th className="border border-slate-300">Nav</th>
                    </tr>
                </thead>
                <tbody>

                {todos.map((todo, index) => (
                        <tr key={index}>
                        <td className="border border-slate-300 p-2"><span className="mr-5">{todo.title}</span></td>
                        <td className="border border-slate-300">                       
                        <Button onClick={() => handleDeleteTodo(todo.id)} className="bg-primary text-white pl-6 pr-6 pt-1 pb-1" text="Delete" testId={`delete-button-${todo.id}`} />
                            </td>
                        </tr>
                    ))}
                </tbody>
            </table>

             <Button onClick={handleClearTodos} className="bg-primary text-white pl-8 pr-8 pt-3 pb-3" text="Clear list"></Button>
        </div>
    )
}

export default TodoList;

