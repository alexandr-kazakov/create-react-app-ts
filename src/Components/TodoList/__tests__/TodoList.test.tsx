import { render, fireEvent, screen } from '@testing-library/react';
import TodoList from '../TodoList';

describe('TodoList component', () => {
  test('renders input and add button', () => {
    render(<TodoList />);
    const inputElement = screen.getByPlaceholderText('Enter your task');
    const addButtonElement = screen.getByText('Add');
    expect(inputElement).toBeInTheDocument();
    expect(addButtonElement).toBeInTheDocument();
  });

  test('adds todo item when add button is clicked', () => {
    render(<TodoList />);
    const inputElement = screen.getByPlaceholderText('Enter your task');
    const addButtonElement = screen.getByText('Add');

    fireEvent.change(inputElement, { target: { value: 'New Todo' } });
    fireEvent.click(addButtonElement);

    const todoItemElement = screen.getByText('New Todo');
    expect(todoItemElement).toBeInTheDocument();
  });

  test('deletes todo item when delete button is clicked', () => {
    const fakeTodo = { id: 1, title: 'Your todo title' };
    render(<TodoList />);
    const deleteButtonElement = screen.getByTestId(`delete-button-${fakeTodo.id}`);
    fireEvent.click(deleteButtonElement);
    expect(screen.queryByText('Your todo title')).toBeNull();
  });
  

    test('clears todo list when clear button is clicked', () => {
        render(<TodoList />);
        const inputElement = screen.getByPlaceholderText('Enter your task');
        const addButtonElement = screen.getByText('Add');

        fireEvent.change(inputElement, { target: { value: 'New Todo 1' } });
        fireEvent.click(addButtonElement);
        fireEvent.change(inputElement, { target: { value: 'New Todo 2' } });
        fireEvent.click(addButtonElement);

        const clearButtonElement = screen.getByText('Clear list');
        fireEvent.click(clearButtonElement);

        // Проверяем, что нет строк с задачами
        const todoItemElements = screen.queryAllByRole('row');
        // Пропускаем первую строку (заголовок таблицы)
        expect(todoItemElements.length - 1).toBe(0);
    });
});


describe('TotoList component snapshot', () => {
    test('renders TodoList component correctly', () => {
        const { container } = render(<TodoList />);
        expect(container).toMatchSnapshot();
    });
});