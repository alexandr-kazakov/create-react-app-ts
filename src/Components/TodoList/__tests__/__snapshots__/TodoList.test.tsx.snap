// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TotoList component snapshot renders TodoList component correctly 1`] = `
<div>
  <div
    class="todo-list"
  >
    <div
      class="mb-10"
    >
      <input
        class="dark:bg-black dark:text-white border border-slate-400 h-12 pl-2 pr-2"
        placeholder="Enter your task"
        type="text"
        value=""
      />
      <button
        class="bg-primary text-white pl-8 pr-8 pt-3 pb-3"
      >
        Add
      </button>
    </div>
    <table
      class="table-auto border-collapse border border-slate-400 max-w-5xl w-full mb-10"
    >
      <thead>
        <tr>
          <th
            class="border border-slate-300"
          >
            Info
          </th>
          <th
            class="border border-slate-300"
          >
            Nav
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <button
      class="bg-primary text-white pl-8 pr-8 pt-3 pb-3"
    >
      Clear list
    </button>
  </div>
</div>
`;
